### ISSUE TITLE
Beds in all wool colours

### ISSUE THREAD

**Issue by paramat** [minetest/minetest_game#1923]
#1794 has some discussion on this, it was closed due to several issues and an unsuitable approach.

I request that beds are made available in all wool colours, instead of just 'MC red'.
This will also solve the often expressed frustration of requiring a rose to make a bed.

1. Provide beds and fancy beds in all MTG wool mod colours, including white (no dye is needed for white wool). Total 30 beds.
2. The parts of the textures that are the coloured bed cover could use 'colourise' or 'multiply' texture modifiers, but not hardware colourisation (static or param2) due to the extra complexity it creates. These textures would then be combined with 'base' textures that are common to all beds.
3. The bed covers need to colour co-ordinate (identical Hue) with the corresponding wools that are probably in use nearby in a house. Our current red bed matches red wool in Hue but has reduced contrast, this should be the case for the additional coloured beds.
4. The bed registration code must support existing bed registrations that are in use in many mods without breaking them.
5. The existing beds should be aliased to the new "beds:bed_red" and "beds:fancy_bed_red".
6. Instead of 30 bed registrations a table of changing parameters should be created and the beds registered in a loop that reads this table.
7. All beds should be present in creative inventory for instant use (obvious but was not present in the closed PR).

**Comment by paramat** [comment]
Updated point 2, wool pixel hues vary greatly so we cannot use 'colourise' or 'multiply' texture modifiers, or hardware colourisation, 15 separate textures are needed for the coloured blankets.

**Comment by hkzorman** [comment]
@paramat I have been trying to figure out what is the contrast and brightness adjustment that has been done from the normal red wool texture to the bed texture. However, while I get close to the same effect, I do notice that the bed texture brightness and contrast are not uniform.
The center seems to be more bright while the edges are darker.
Is it wished for this pattern to continue? In my opinion, it doesn't make sense. I'm talking here about the `beds_bed_top_bottom.png` file. This is part of a blanket, and therefore having a distinct square with edges at the bottom of the bed doesn't seems good to me, I think a uniform brightness and contrast adjustment is better.

**Comment by paramat** [comment]
Close is good enough.

I just tiled the texture and yes the edges have been darkened which is not good or at best unnecessary, uniform is fine.

**Comment by hkzorman** [comment]
Awesome, thanks for clarification!

**Comment by paramat** [comment]
Any part of the textures which is coloured should be uniform, no edge darkening.

**Comment by hkzorman** [comment]
Here is an example. The left texture is the original, `beds_bed_top_bottom`. The right texture is my texture.

![textures](https://user-images.githubusercontent.com/24514440/32276050-95c159e4-bedb-11e7-87de-573858165802.png)

Is this coloring ok?

**Comment by DS-Minetest** [comment]
Without edge darkening the texture looks a bit flat.

**Comment by paramat** [comment]
> Here is an example

Perfect.

DS-Minetest it has to be uniform as it part of a larger area of coloured blanket.

**Comment by paramat** [comment]
@hkzorman all coloured blankets should have that lower contrast of the red one, but they don't all need to be darkened by the same amount. For example already dar colours won't need darkening and some colours may not look as godd darkened (such as yellow) so i'll leave that to your artistic judgement.

**Comment by hkzorman** [comment]
@paramat These are all the wool textures modified for bed usage. I will use these base textures in all bed textures that require color. Are these ok to you?
![base_black_black](https://user-images.githubusercontent.com/24514440/32497415-7964f3c8-c39a-11e7-837c-a9635404b4b7.png)
![wool_blue](https://user-images.githubusercontent.com/24514440/32497422-7ec0504c-c39a-11e7-9c54-b0a9f197cda8.png)
![wool_brown](https://user-images.githubusercontent.com/24514440/32497425-82952666-c39a-11e7-9ae5-30a173b11976.png)
![wool_cyan](https://user-images.githubusercontent.com/24514440/32497441-9837e49a-c39a-11e7-9465-8eefdb813964.png)
![wool_dark_green](https://user-images.githubusercontent.com/24514440/32497452-a31f6982-c39a-11e7-949e-13e8443c3e54.png)
![wool_dark_grey](https://user-images.githubusercontent.com/24514440/32497466-a63e0088-c39a-11e7-8d34-866b53960751.png)
![wool_green](https://user-images.githubusercontent.com/24514440/32497477-ab452e1c-c39a-11e7-8384-b4bab013d399.png)
![wool_grey](https://user-images.githubusercontent.com/24514440/32497481-af3ce604-c39a-11e7-82ff-fcedd83f071a.png)
![wool_magenta](https://user-images.githubusercontent.com/24514440/32497500-b9f620b0-c39a-11e7-951a-c0a0dc5d1bf1.png)
![wool_orange](https://user-images.githubusercontent.com/24514440/32497511-c256d574-c39a-11e7-80dd-ce7da026c383.png)
![wool_pink](https://user-images.githubusercontent.com/24514440/32497514-c82caa64-c39a-11e7-8210-c12a7b642e7a.png)
![base_red](https://user-images.githubusercontent.com/24514440/32497522-cd0b3474-c39a-11e7-94de-f92bcb6ebcd7.png)
![wool_violet](https://user-images.githubusercontent.com/24514440/32497530-d2cb88fa-c39a-11e7-9186-e34d5ac2f8c6.png)
![wool_white](https://user-images.githubusercontent.com/24514440/32497540-db3fc4e2-c39a-11e7-8219-7722eaf74ab8.png)
![wool_yellow](https://user-images.githubusercontent.com/24514440/32497543-ddedf484-c39a-11e7-8c08-0a92dfa23947.png)

**Comment by DS-Minetest** [comment]
The light green and the orange could be brighter imho.

**Comment by paramat** [comment]
I'll check them in-game.

**Comment by hkzorman** [comment]
I'm already half-way modifying the textures, so I won't proceed further until you are ok with them.

**Comment by paramat** [comment]
Looking at the small images they seem ok but i will compare against the wools in Gimp too.

**Comment by paramat** [comment]
Just looked at black and blue so far.
The average hue varies between your bed blue and the wool blue.
If you open both in Gimp, select the 'colour picker' tool, hold shift and click a pixel, a box opens where you can choose to show HSV values.
Now if you hold down left mouse button and drag the pointer over the pixels you can get an idea of the hue variation. Your texture is Hue 216-217 and the wool is Hue 208-213.

I tried reducing contrast of the wool texture to see how the process is done, and i can see it's not easy to preserve the hue of each pixel, what process did you use?
If it's too difficult perhaps the texture will have to be colourised to one hue only, not sure.

**Comment by paramat** [comment]
In trying to do the process myself i can see it is difficult. I started with reducing contrast, then used 'HSV' tool too boost saturation multiple times, then i needed to reduce contrast again, boost saturation again.

I think it's more important to have hue matching than to reproduce the reduction of contrast. Perhaps we can use the wool colours as they are or just make them darker, no reduction of contrast?

**Comment by paramat** [comment]
Let's get another opinion on this.
@rubenwardy @SmallJoker what do you think about bed colours being exact copies of the wool textures? I actually don't like how the red bed blanket doesn't match any wool textures used nearby.

**Comment by paramat** [comment]
Just looked at the brown. It's an unpleasant oversaturated red-brown now, pixel hues are very different.
Cyan has non-matching hues but looks ok.
Dark green: Wool is H99-107, your texture is 120 for all.
Dark grey is ok.
Green, wool H97-106, yours 120 for all.
Grey ok.
Magenta has hue mismatch.
Orange has hue mismatch.

I won't do the others.

**Comment by paramat** [comment]
Other devs are ok with exactly matching the wool textures http://irc.minetest.net/minetest-dev/2017-11-10#i_5132745

**Comment by hkzorman** [comment]
So, no texture modification, just use the same wool texture for all colors?

**Comment by paramat** [comment]
For each bed colour, simply use the corresponding wool texture unchanged.
And sorry for the change in direction.

**Comment by hkzorman** [comment]
Ok, no problem, I will do so.
I had a basic formula for most of them, except the very dark and very bright textures. It was this:
* Saturation: 150
* Lightness: -25
* Contrast: -100
* Brightness: -85

For the darkest ones, I applied around -25 brightness and -25 contrast and nothing else.

No problem for the change of direction, I understand that some of my textures weren't of the best colors! I will work on these now.

**Comment by paramat** [comment]
Yes in trying to process one myself it seems that the contrast change can change hue values, i had not expected this so sorry for that suggestion.

**Comment by hkzorman** [comment]
@paramat No problem, I should be able to come up with a PR within this week.

**Comment by paramat** [comment]
See https://github.com/minetest/minetest_game/blob/master/mods/flowers/init.lua#L29 for how to register large numbers of nodes by iterating through a data table. The registration API of beds can remain the same, then alias the old beds to the new red bed.

**Comment by paramat** [comment]
Looks good.

**Comment by 4w** [comment]
Instead of registering nodes for beds in all colors maybe use hardware colors instead?

**Comment by paramat** [comment]
See point 2 in first post.

**Comment by 4w** [comment]
> See point 2 in first post.

See what I quoted and read what I replied to that quote.

**Comment by paramat** [comment]
If you originally quoted point 2 in your comment, sorry.
By 'optimise the textures' do you suggest changing the textures so that they have all pixels at the same hue and using static hardware colouring? it's not very clear.

If so, wool textures would have to be changed too as the intent is to have beds match wool. However, our wool textures look good because they have varying hue pixels. The problem with hardware colouring is that it forces you to use a single base texture and applies a single hue to it, this removes a lot of possibilities and results in poorer looking textures.

**Comment by 4w** [comment]
> If so, wool textures would have to be changed too as the intent is to have beds match wool.

Yes. Change wool and beds to use hardware coloring exclusively and optimize the textures to look good in all colors.

> this removes a lot of possibilities and results in poorer looking textures.

*[citation needed]*

**Comment by paramat** [comment]
Would also break all texture packs and alter the appearence of builds.
Hardware colouring textures always look worse to me due to their limitations.

**Comment by 4w** [comment]
> Would also break all texture packs and alter the appearence of builds.

0.5.0 is a breaking release, so … :smile:

**Comment by paramat** [comment]
The limitation is how they inherently work, a base texture with a colour applied, that's unavoidable.

**Comment by sfan5** [comment]
There is no stability guarantee for texture packs, they have been broken in the past.
This is not an argument.

(Same goes for "appearance of builds" by the way.)

**Comment by paramat** [comment]
True, but '0.5 breakage' still doesn't mean breaking texture packs.
The appearence comment is just because i expect it would be unpopular to reduce the quality of wool textures.

**Comment by paramat** [comment]
@hkzorman any progress?

**Comment by hkzorman** [comment]
@paramat apologies for the long delay. I took long time to edit the inventory icons to look acceptable.
Hopefully they are now, at any rate, I opened a PR and then we can change the icons if needed.
PR #2052

**Comment by paramat** [comment]
No problem, 0.5 is not soon.

**Comment by beyondlimits** [comment]
IMO the requirement of a rose (when there was only one color of bed anyway) was a poor, shameful mistake, not a feature.

**Comment by paramat** [comment]
Yes i agree. It tried to mimic MC too closely.
